### PR TITLE
ts/MACD-indicator

### PR DESCRIPTION
**Pull request by K009**

### CODE REVIEW

File: js/masters/indicators/indicators-all.src.js [hunk 1]
```diff
@@ -30,7 +30,8 @@ import '../../Stock/Indicators/TRIXIndicator.js';
 import '../../Stock/Indicators/APO/APOIndicator.js';
 import '../../Stock/Indicators/IKHIndicator.js';
 import '../../Stock/Indicators/KeltnerChannelsIndicator.js';
-import '../../Stock/Indicators/MACDIndicator.js';
+import '../../Stock/Indicators/MACD/MACDIndicator.js';
+import '../../Stock/Indicators/MFIIndicator.js';
 import '../../Stock/Indicators/MFI/MFIIndicator.js';
```
Comment (bre1470) [inline js/masters/indicators/indicators-all.src.js:35]: Solving the conflict failed.

File: js/Stock/Indicators/MACDIndicator.js [hunk 2]
```diff
@@ -6,15 +6,31 @@
...
+            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
+        return extendStatics(d, b);
+    };
+    return function (d, b) {
+        extendStatics(d, b);
+        function __() { this.constructor = d; }
+        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
+    };
+})();
+var _a = BaseSeries.seriesTypes, SMAIndicator = _a.sma, EMAIndicator = _a.ema;
```
Comment (bre1470) [inline js/Stock/Indicators/MACDIndicator.js:22]: This whole file should be deleted.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+    options: Highcharts.MACDIndicatorOptions;
+    parallelArrays: Array<string>;
+    pointArrayMap: Array<string>;
+    pointClass: typeof Highcharts.MACDIndicatorPoint;
+    points: Array<Highcharts.MACDIndicatorPoint>;
+    pointValKey: string;
+    requiredIndicators: Array<string>;
+    signalZones: Highcharts.MACDIndicatorZonesOptions;
+    toYData(point: Point): Array<number>;
+    translate(): void;
```
Comment (bre1470) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:200]: Regular properties (not mentioned in extend call, excepted pointClass) and functions got into the class pattern.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+    parallelArrays: ['x', 'y', 'signal', 'MACD'],
+    pointValKey: 'y',
+    // Columns support:
+    markerAttribs: (noop as any),
+    getColumnMetrics: H.seriesTypes.column.prototype.getColumnMetrics,
+    crispCol: H.seriesTypes.column.prototype.crispCol,
+    // Colors and lines:
+    init: function (
+        this: MACDIndicator
+    ): void {
```
Comment (bre1470) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:217]: Regular functions go into the class pattern.

**Review by bre1470 — CHANGES_REQUESTED** [review]
Thank you! Please see requested changes.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+                j++;
+            }
+        }
+
+        return {
+            values: MACD,
+            xData: xMACD,
+            yData: yMACD
+        } as IndicatorValuesObject<TLinkedSeries>;
+    }
```
Comment (bre1470) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:477]: Functions should be set explicitly public.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+    graphmacd: (Highcharts.SVGElement|undefined);
+    graphsignal: (Highcharts.SVGElement|undefined);
+    macdZones: MACDZonesOptions;
+    nameComponents: Array<string>;
+    parallelArrays: Array<string>;
+    pointArrayMap: Array<string>;
+    pointClass: typeof MACDPoint;
+    pointValKey: string;
+    requiredIndicators: Array<string>;
+    signalZones: MACDZonesOptions;
```
Comment (bre1470) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:493]: The following properties are not part of the extend call and should be move to the class pattern: currentLineZone, graphmacd, graphsignal, macdZones, signalZones.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+            yData: yMACD
+        } as IndicatorValuesObject<TLinkedSeries>;
+    }
+}
+
+interface MACDIndicator {
+    crispCol: typeof ColumnSeries.prototype['crispCol'];
+    currentLineZone: (string|null);
+    getColumnMetrics: typeof ColumnSeries.prototype['getColumnMetrics'];
+    graphmacd: (Highcharts.SVGElement|undefined);
```
Comment (bre1470) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:484]: Avoid references of the internal Highcharts namespace. Make use of a type import for SVGElement.

**Review by bre1470 — CHANGES_REQUESTED** [review]
Thank you! Please see comments for requested changes.

File: ts/Stock/Indicators/MACD/MACDIndicator.ts [hunk 8]
```diff
@@ -479,18 +485,13 @@ class MACDIndicator extends SMAIndicator {
...
 interface MACDIndicator {
     crispCol: typeof ColumnSeries.prototype['crispCol'];
-    currentLineZone: (string|null);
     getColumnMetrics: typeof ColumnSeries.prototype['getColumnMetrics'];
-    graphmacd: (Highcharts.SVGElement|undefined);
-    graphsignal: (Highcharts.SVGElement|undefined);
-    macdZones: MACDZonesOptions;
     nameComponents: Array<string>;
     parallelArrays: Array<string>;
     pointArrayMap: Array<string>;
```
Comment (sebastianbochan) [inline ts/Stock/Indicators/MACD/MACDIndicator.ts:491]: `extend()` includes params, so you do not need them in the interface. TD:LD: `parallelArrays, pointValKey` etc should be removed here.

**Review by sebastianbochan — CHANGES_REQUESTED** [review]

File: ts/Stock/Indicators/MACD/MACDIndicator.ts (hunk 8 above)
Comment (bre1470) [reply ts/Stock/Indicators/MACD/MACDIndicator.ts:491]: Please do not remove them, if they were defined on the class in the internal namespace. As long as they are in the extend call, they should also be in the interface.

Comment (sebastianbochan) [reply ts/Stock/Indicators/MACD/MACDIndicator.ts:491]: I was pretty sure that its enough in the extend only, my mistake.

Comment (bre1470) [reply ts/Stock/Indicators/MACD/MACDIndicator.ts:491]: Unfortunately the extend call is not type safe.

**Review by bre1470 — APPROVED** [review]
Thank you! Looks good. 👍

**Review by bre1470 — APPROVED** [review]

**Pull request merged**